### ISSUE TITLE
Gradient Overflow Warnings when using the Dice Loss

### ISSUE THREAD

**Issue by vgoklani** [NVIDIA/apex#592]
Hey there,

I'm getting Gradient Overflow warnings when using the Dice Loss and with apex/amp set to O1/O2. How should I fix this?

	import torch

	class DiceLoss(torch.nn.Module):
		""" https://lars76.github.io/neural-networks/object-detection/losses-for-segmentation/
		"""
		def __init__(self, number_classes, weight=None):
			super().__init__()
			self._number_classes = number_classes
			self._weight = weight

		@staticmethod
		def calculate_dice_loss(prediction, target, smooth: float = 1):
			prediction_flat = torch.flatten(prediction)
			target_flat = torch.flatten(target)

			numerator = 2 * torch.sum(prediction_flat * target_flat) + smooth
			denominator = torch.sum(prediction_flat + target_flat) + smooth

			return 1 - (numerator / denominator)

		def forward(self, predictions, target):
			totalLoss = 0
			for i in range(self._number_classes):
				diceLoss = DiceLoss.calculate_dice_loss(predictions[:, i], target)
				if self._weight is not None:
					diceLoss *= self._weight[i]
				totalLoss += diceLoss
			return totalLoss

Thanks!

**Comment by mcarilli** [comment]
Getting gradient overflow warnings for the first few (5-10) iterations is normal, as the value used to scale the gradients calibrates.  It's also normal to see the warning intermittently (every few hundred or thousand steps afterwards) as the scale occasionally adjusts.

A death spiral (dozens of iterations where the scale value decreases to very small values, eventually resulting by division by zero) is not normal, and indicates divergence.  Please reopen if that's what you observe.
